google-account-plugin/src/com/google/cloud/tools/intellij/stats/UsageTracker.java

```diff
@@ -33,4 +34,26 @@ void trackEvent(@NotNull String eventCategory,
       @Nullable String eventLabel,
       @Nullable Integer eventValue);
 
+  PartialTrackingEventAction trackEvent(String category);
+
+  interface SendsEvent {
+
+    void send();
+  }
+
+
+  interface PartialTrackingEventAction extends SendsEvent {
+
+    PartialTrackingEventLabel withAction(String value);
+
+    interface PartialTrackingEventLabel extends SendsEvent {
+
+      PartialTrackingEventValue andLabel(String value);
+
+      interface PartialTrackingEventValue extends SendsEvent {
```

Comment:
It's trying to be fluent, and maybe the scoping of the interfaces is
gratuitous. I wouldn't be super opposed to breaking them up as siblings. I
definitely would have done that if I thought there were use-cases where the
client might actually store one of these partial types as a left-hand-side
value.

On Wed, Jun 22, 2016 at 2:05 PM, Etan Shaul <notifications@github.com>
wrote:

> In
> google-account-plugin/src/com/google/cloud/tools/intellij/stats/UsageTracker.java
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/752#discussion_r68104191>
> :
>
> > +
> > +  interface SendsEvent {
> > +
> > +    void send();
> > +  }
> > +
> > +
> > +  interface PartialTrackingEventAction extends SendsEvent {
> > +
> > +    PartialTrackingEventLabel withAction(String value);
> > +
> > +    interface PartialTrackingEventLabel extends SendsEvent {
> > +
> > +      PartialTrackingEventValue andLabel(String value);
> > +
> > +      interface PartialTrackingEventValue extends SendsEvent {
>
> the Sequential Builder Pattern? I just made that up :)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/752/files/baccf888600f472b69946034e2bc213cc9ce67ec#r68104191>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AHf5HZZ2jiEU-f1qyESGf59IboacWqtgks5qOXl-gaJpZM4I8B5a>
> .
>